---
title: "Codigos_Organizados"
author: "Wasim Syed"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Pacotes recomendados

### CRAN packages

```{r eval=FALSE, include=FALSE, echo =FALSE}
#CRAN packages ----------------
# install.packages("ggbeeswarm")
# install.packages("ggrepel")
# install.packages("tidyverse")
# install.packages("org.Hs.eg.db")
# install.packages("RColorBrewer")
# install.packages("plotly")
# install.packages("msigdbr")
# install.packages("ape")
# install.packages("sva")
# install.packages("gplots")
# install.packages("tibble")
# install.packages("metaRNASeq")
# install.packages("editData")
# install.packages("tidyverse")
# install.packages("corrr")
# install.packages("ggcorrplot")
# install.packages("FactoMineR")
# install.packages("factoextra")
# install.packages("esquisse")
# install.packages("corto")
# install.packages("reshape2")
# install.packages("explore")
# install.packages("extrafont")
# install.packages("showtext")
# install.packages("flexdashboard")
# install.packages("ggheatmap")
# install_github("jokergoo/InteractiveComplexHeatmap")
```

### Bioconductor packages
```{r eval=FALSE, include=FALSE, echo =FALSE}
# Bioconductor packages ------------------------
# install.packages("BiocManager") 
# BiocManager::install("GEOquery") 
# BiocManager::install("org.Hs.eg.db")
# BiocManager::install("deseq2")
# BiocManager::install("biomaRt")
# BiocManager::install("celldex")
# BiocManager::install("NMF")
# BiocManager::install("BiasedUrn")
# BiocManager::install("GSVA")
# BiocManager::install("sva")
# BiocManager::install("clusterProfiler")
# BiocManager::install("ggupset")
# BiocManager::install("msigdbr")
# BiocManager::install("EnhancedVolcano")
# BiocManager::install("AnnotationDbi")
```

### Call packages
```{r eval=FALSE, include=FALSE, echo =FALSE}
# Call packages -----------------
library(devtools)
library(GEOquery)
library(Matrix)
library(circlize)
library(RColorBrewer)
library(celldex)
library(biomaRt)
library(org.Hs.eg.db)
library(plotly)
library(DESeq2)
library(msigdbr)
library(ape)
library(GSVA)
library(sva)
library(clusterProfiler)
library(EnhancedVolcano)
library(ggbeeswarm)
library(corrr)
library(ggcorrplot)
library(FactoMineR)
library(factoextra)
library(esquisse)
library(ComplexHeatmap)
library(janitor)
library(ggsci)
library(patchwork)
library(ggthemes)
library(here)
library(extrafont)
library(showtext)
library(readxl)
library(scales)
library(tidyverse)
library(flexdashboard)
library(ggheatmap)
library(vegan)
library(InteractiveComplexHeatmap)

```

# Customize themes
### Create your own theme
```{r}
# Custom theme
theme_vaxgo <- function() {
  theme_minimal() +
  theme(
    text = element_text(color = "black", 
                        size = 12,
                        family = "arial",
                        hjust = -1),
    plot.title = element_text(family = "arial",
                              size = 16,
                              face = "bold"),
    plot.subtitle = element_text(family = "arial",
                                 size = 12),
    axis.text.y = element_text(hjust = 1,
                               color = "black"),
    axis.text.x = element_text(hjust = 0.5,
                               color = "black"),
    axis.title = element_text(size = 12),
    axis.title.x = element_text(hjust = 0.5),
    legend.title = element_text(face = "bold",
                                size = 10),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.4, 'cm'),
    axis.ticks = element_line(linewidth = 0.5),
    plot.background = element_rect(fill='white', colour='white'),
    plot.caption = element_text(hjust = 0,
                                family = "arial"),
    panel.grid = element_line(linewidth = 0.1,
                              color = "grey97")
  )
}
```

### Custom palettes
```{r}

# For a single categoric variable
custom_colors = c("Category1" = "#bc3908",
                  "Category2" = "#e76f51")

# For different variables in a dataframe
list_custom_colors = list(
  custom_colors_variable1 = c("Category1" = "#bc3908",
                              "Category2" = "#e76f51"),
  custom_colors_variable2 = c("0" = "#ECF8FA",
                              "1" = "#caf0f8",
                              "2" = "#6CD5EA",
                              "3" = "#0087BF",
                              "4" = "#015BA0",
                              "6" = "#03045e",
                              "7" = "#03045e",
                              "8" = "#03045e"),
  custom_colors_variable3 = c("0" = "grey95",
                              "1" = "#91D1C2A0",
                              "2" = "#00A087FF",
                              "3" = "#3C5488FF"))

```

### Flourish
```{r}
#Colors for Flourish
formatted_vector <- paste(names(col_immune), ":", col_immune) %>% 
  cat(., sep = "\n")
```


### Scientific palettes (ggsci)
```{r}

# Palette JAMA
# pal_jama("default")(10) %>% 
#   show_col()


# Palette COSMIC
# pal_cosmic("hallmarks_light")(10) %>%
#   show_col()
```



# Bioinformatics


## Calculate auto overlap (Conditions)
```{r}
############# Input

#Files
conditions_genes = read_csv(here("Example", "all_degs_p_05_vac_infected_19_12_23.csv"))
annotation_conditions = read_csv(here("Example", "ann_vaccines_conditions.csv"), 
    col_types = cols(...1 = col_skip()))

# Select columns for annotation
columns_annotation = c("condition", "day", "week", "dose", "type", "disease_vac")


#Name your outputs
filename = "my_condition_degs"

#Filters
filter_padj = 0.05
filter_logfc = 1

#DEGs table 

# Only significant
degs_sig = conditions_genes %>% 
  filter(padj < filter_padj,
         log2fold_change > filter_logfc | log2fold_change < -filter_logfc)

# Only up and down
data = conditions_genes %>% 
  filter(padj < filter_padj,
         log2fold_change > filter_logfc | log2fold_change < -filter_logfc) %>% 
  select(process = condition, genes)
```

Function for overlapping 
```{r}
############### VENN DIAGRAM

# Function for overlapping -------
overlap_genes <- function(cond1, cond2, data) {
  genes_cond1 <- data$genes[data$process == cond1]
  genes_cond2 <- data$genes[data$process == cond2]
  
  genes_shared <- intersect(genes_cond1, genes_cond2)
  
  genes_notshared_cond1 <- setdiff(genes_cond1, genes_cond2)
  genes_notshared_cond2 <- setdiff(genes_cond2, genes_cond1)
  
  total_genes_cond1 <- length(genes_cond1)
  total_genes_cond2 <- length(genes_cond2)
  
  percentage_shared_cond1 <- length(genes_shared) / total_genes_cond1 * 100
  percentage_shared_cond2 <- length(genes_shared) / total_genes_cond2 * 100
  
  shared_genes <- data.frame(
    Cond1 = cond1,
    Cond2 = cond2,
    Shared = length(genes_shared),
    NotShared_cond1 = length(genes_notshared_cond1),
    NotShared_cond2 = length(genes_notshared_cond2),
    Total_Genes_Cond1 = total_genes_cond1,
    Total_Genes_Cond2 = total_genes_cond2,
    Genes_Names = paste(genes_shared, collapse = ", "),
    Percentage_Shared_Cond1 = percentage_shared_cond1,
    Percentage_Shared_Cond2 = percentage_shared_cond2
  )
  
  return(shared_genes)
}
```

Perform calculations
```{r}
# Perform calculations
unique_cond <- unique(data$process) # List sets
shared_genes_df <- data.frame() # Store data

for (i in 1:(length(unique_cond) - 1)) {
  for (j in (i + 1):length(unique_cond)) {
    resultado_temp <- overlap_genes(unique_cond[i], unique_cond[j], data)
    shared_genes_df <- bind_rows(shared_genes_df, resultado_temp)
  }
}

# Fisher's exact test
fisher_exact_test <- function(shared, notshared1, notshared2) {
  cont_table <- matrix(c(shared, notshared1, notshared2, 0), nrow = 2)
  results_fisher <- fisher.test(cont_table)
  return(results_fisher$p.value)
}
shared_genes_df = shared_genes_df %>%
  mutate(pvalue = pmap_dbl(list(Shared, NotShared_cond1, NotShared_cond2), fisher_exact_test)) %>% 
  mutate("-log(p)" = -log10(pvalue))


#Mirror table
shared_genes_df2 = shared_genes_df %>% 
  rename(Cond1_temp = Cond1, Cond2_temp = Cond2) %>%
  rename(Cond1 = Cond2_temp, Cond2 = Cond1_temp)

shared_genes_df_mirror = shared_genes_df %>% 
  bind_rows(shared_genes_df2) %>% 
  mutate(Percentage_Shared_Cond1 = round(Percentage_Shared_Cond1, 2),
         Percentage_Shared_Cond2 = round(Percentage_Shared_Cond2, 2))
  
#Jaccard distance
jaccard.matrix <- data %>%
  mutate(present = 1) %>% 
  distinct() %>% 
  pivot_wider(names_from = genes, values_from = present, values_fill = 0) %>%
  column_to_rownames(var = "process") %>%
  vegdist(binary = TRUE, method = "jaccard", diag = TRUE, upper = TRUE) %>%
  as.matrix() %>%
  {1 - .}

shared_genes_df_mirror = jaccard.matrix %>% 
  as.data.frame() %>% 
  rownames_to_column("Cond1") %>% 
  pivot_longer(cols = -"Cond1",
               names_to = "Cond2",
               values_to = "jaccard_distance") %>% 
  inner_join(shared_genes_df_mirror, by = join_by(Cond1, Cond2)) %>% 
  mutate(jaccard_distance = round(jaccard_distance, 3))

#Save
write_csv(shared_genes_df_mirror, 
          file = here("Example", paste0(filename, "_sharedgenes_Fisher_Jaccard.csv")))
```

### Flourish Circos plot from overlapping genes

```{r}

#Filters 
filter_shared = 1
filter_pvalue = 0.10

# Create table for flourish
shared_genes_df_mirror %>% 
  filter(Shared > filter_shared,
         pvalue < filter_pvalue) %>% 
  select(Cond1, Cond2, Shared, Total_Genes_Cond1, Percentage_Shared_Cond1, pvalue:"-log(p)") %>% 
  write_csv(file = here("Example",  paste0(filename,"_Flourish_Circos_Links.csv")))


#Define colors for flourish
color_flourish = "type"
ann_flourish = annotation_conditions %>% 
  select(type) 

#Colors for Flourish
colors = list_custom_colors$color_flourish %>%
  as.data.frame() %>%
  rownames_to_column(color_flourish) %>%
  mutate(color_flourish = str_to_sentence(color_flourish)) %>% 
  rename(color = ".")

colors_flourish_table = ann_flourish %>%
  clean_names() %>%
  select(!color_flourish) %>%
  inner_join(colors, by = color_flourish)

paste(colors_flourish_table$condition, ":", colors_flourish_table$color_flourish) %>%
  cat(., sep = "\n")

```


### Heatmap (overlap counts)

```{r}

filter_pvalue_heat = 0.10

matrix_data = shared_genes_df_mirror %>% 
  filter(pvalue <= filter_pvalue_heat) %>% 
  distinct() %>% 
  select(Cond1, Cond2, Shared) %>% 
  pivot_wider(names_from = "Cond1",
              values_from = "Shared") %>% 
  column_to_rownames("Cond2") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

# Column and row annotation
######## Annotations
annotation_heatmap = annotation_conditions %>% 
  select(columns_annotation) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  rename(condition = '.') %>% 
  inner_join(annotation_heatmap, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") 
  # mutate(immune_system = str_to_upper(immune_system)) %>% 
  # select(immune_system)

matrix_data = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(immune_system)) %>% 
  arrange(module_title) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Rows Immune ----
ann_rows_heatmap = ann_cols_heatmap

mat = matrix_data %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "module_title") %>% 
  inner_join(ann_rows_heatmap %>% rownames_to_column("module_title"), by = "module_title") %>% 
  select(!c(immune_system)) %>% 
  arrange(module_title) %>% 
  column_to_rownames("module_title") %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Check dimensions
dim(mat)
dim(ann_cols_heatmap)
dim(ann_rows_heatmap)

colors_immune = list(immune_system = col_immune)

```


```{r fig.width= 10, fig.height=8}
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = colors_immune,
                       annotation_name_side = "right")

row_ha = row_ha = HeatmapAnnotation(df = ann_rows_heatmap,
                       col = colors_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F)

# Values colors
col_fun <- colorRamp2(c(0, 100), c("white", "#ed6a5a"))

file = "BTM_Modules_Overlapping_ImmuneSystem"

width_image = 20
height_image = 20
width = unit(10, "cm")
height = unit(10, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = here("Figures", paste0(file, sep ="_", "Complexheatmap.png"))

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        # #cell_fun = function(j, i, x, y, width, height, fill) {
                        #                     if(mat[i, j] > 0)
                        #                     grid.text(sprintf("%.f", mat[i, j]), 
                        #                               x, y, 
                        #                               gp = gpar(fontsize = 5))},
                        show_column_names = TRUE,
                        column_title = file,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "black",
                        name = "Shared",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "bottom")

heatmap_plot

dev.off()


```

### Heatmap (Jaccard distance)

```{r fig.width= 10, fig.height=8}

mat = jaccard.matrix

#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = colors_immune,
                       annotation_name_side = "right")

row_ha = row_ha = HeatmapAnnotation(df = ann_rows_heatmap,
                       col = colors_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F)

# Values colors

max_value = mat %>% 
   replace(. == 1.00000, 0) %>% 
   max()

col_fun <- colorRamp2(c(0, max_value), c("white", "#ed6a5a"))

file = "BTM_Modules_Jaccard_ImmuneSystem_"
width_image = 20
height_image = 20
width = unit(10, "cm")
height = unit(10, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = here("Figures", paste0(file, sep ="_", "Complexheatmap.png"))

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        cell_fun = function(j, i, x, y, width, height, fill) {
                                            if(mat[i, j] > 0)
                                            grid.text(sprintf("%.2f", mat[i, j]), 
                                                      x, y, 
                                                      gp = gpar(fontsize = 5))},
                        show_column_names = TRUE,
                        column_title = file,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "black",
                        name = "Jaccard distance",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(1, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = T,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "bottom")

dev.off()

heatmap_plot

```



## Gene set enrichment analysis (ORA and GSEA)

### Prepare gene sets
```{r}
########## Gene sets ------

######## IMMUNEGO
ImmuneGO_Annotated_GO <- read_csv(here("Tables", "ImmuneGO_Annotated_GO_2024-05-28.csv"))

ImmuneGO_annotation = ImmuneGO_Annotated_GO %>% 
  select(process,immune_system:immune_tissue) 

ImmuneGO_genes_general = ImmuneGO_Annotated_GO %>% 
  separate_rows(genes, sep = ",") %>% 
  filter(go_term == "Manual",
         !process %in% c("ADAPTIVE IMMUNE SYSTEM",
                         "INNATE IMMUNE SYSTEM",
                         "BCR REPERTOIRE",
                         "TCR REPERTOIRE")) %>% 
  select(process, genes)

ImmuneGO_genes_specific = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("ADAPTIVE IMMUNE SYSTEM",
                         "INNATE IMMUNE SYSTEM",
                         "BCR REPERTOIRE",
                         "TCR REPERTOIRE")) %>% 
  select(process=gene_set_short, genes)

######## CELL MARKER IMMUNE ---------
CellMarker_annotation = CellMarker_ImmuneCells %>% 
  select(cell_name, Type) %>% 
  distinct()

CellMarker_genes = CellMarker_ImmuneCells %>% 
  select(process = cell_name, genes = marker)

######## VAX MSIGDB --------
VaxSigDB_Gene_sets_Annotated_RAW <- read_csv(here("Tables", "VaxSigDB_Gene_sets_Annotated_RAW.csv"))

#Filters
vaxsig_columns = c("VACCINE", "PLATFORM", "TARGET_PATHOGEN_DISEASE", "MICROBE_TYPE", "STUDY_TYPE", "STUDY_SUBTYPE","DATE-TIME", "DATE", "TIME", "AGE", "AGE_CATEGORY", "SAMPLE_SOURCE", "SYSTEMATIC_NAME", "GENE_SYMBOLS")
filter_vaxsig_samplesource = "PBMC"
filter_vaxsig_studytype = "VACCINE"
filter_vaxsig_studysubtype = "VAC ONLY"

VaxSigDB_Genesets_filtered = VaxSigDB_Gene_sets_Annotated_RAW %>% 
  select(vaxsig_columns) %>% 
  filter(SAMPLE_SOURCE == filter_vaxsig_samplesource,
         STUDY_TYPE == filter_vaxsig_studytype,
         STUDY_SUBTYPE == filter_vaxsig_studysubtype) 
#Annotation
VaxSigDB_annotation = VaxSigDB_Genesets_filtered %>% 
  mutate(process = paste0(VACCINE, " (", `DATE-TIME`, ", ", AGE, " YO)")) %>% 
  select(process, PLATFORM, TARGET_PATHOGEN_DISEASE, MICROBE_TYPE, DATE, TIME)
#Genes
VaxSigDB_Genes_filtered = VaxSigDB_Genesets_filtered %>% 
  separate_rows(GENE_SYMBOLS, sep = ",") %>% 
  mutate(process = paste0(VACCINE, " (", `DATE-TIME`, ", ", AGE, " YO)")) %>% 
  select(process, genes = GENE_SYMBOLS)

######## BTM MODULES --------
btm_annotation_table <- read_csv(here("Tables", "btm_annotation_immune_table_labelled.csv"))
filter_btm_category = "immune"

#Filters
filter_btm_annotationlevel = c("complete", "partial") #Anyone
filter_btm_annotationlevel = "complete"
filter_btm_annotationlevel = "partial"
#Annotation
btm_annotation = btm_annotation_table %>% 
  filter(module_category == filter_btm_category,
         annotation_level %in% filter_btm_annotationlevel) %>% 
  select(immune_system, immune_subsystem, cells, composite_name)
#Genes
btm_genes = btm_annotation_table %>% 
  filter(module_category == filter_btm_category,
         annotation_level %in% filter_btm_annotationlevel) %>% 
  select(composite_name, module_member_genes) %>% 
  separate_rows(module_member_genes, sep = ",")
```


### Input

```{r}
#Tables

VaxSigDB_annotation
VaxSigDB_Genes_filtered
btm_annotation
btm_genes
CellMarker_annotation
CellMarker_genes
ImmuneGO_genes_general
ImmuneGO_genes_specific
annotation_conditions

######### INPUT HERE ######### 
# Data
degs = degs_sig  %>% 
  filter(condition == "BNT (V2, D1)")


organism = "org.Hs.eg.db"
BiocManager::install(organism, character.only = TRUE)
library(organism, character.only = TRUE)

# Select gene set
TERM2GENE = ImmuneGO_genes_general
geneset_name = "ImmuneGO_genes_general"

```


### Over-representation analysis (ORA)

#### General Gene ontologies
```{r}
# Filter non-neutral DEGs
df <- degs %>% 
  filter(direction != "NEUTRAL")

ora_results <- list()

for (condition_2 in unique(df$condition)) {

    genes <- df %>% 
    filter(condition == condition_2) %>% 
    pull(genes)

  go_enrich <- tryCatch({
    enrichGO(gene = genes,
             OrgDb = organism, 
             keyType = 'SYMBOL',
             readable = T,
             ont = "BP",
             pvalueCutoff = 0.05, 
             qvalueCutoff = 0.10) %>% 
      as.data.frame() %>%
      mutate(condition = condition_2)
  }, error = function(e) {
    return(NULL)
  })

  if (!is.null(go_enrich)) {
    ora_results[[condition_2]] <- go_enrich   
  }
}

ora_results_final <- bind_rows(ora_results)

write.csv(ora_results_final, file = here("Example", paste0(filename, "allGOs_ORA.csv")), row.names = F)


#Filter ancestor GOs
go_bp_ancestor =  GOBPANCESTOR %>% 
  as.data.frame()
names(go_bp_ancestor[, 1]) = "go_child"

go_filtered = ora_results_final %>% 
  clean_names() %>% 
  rownames_to_column("delete") %>% 
  mutate(log_fdr = -log10(qvalue)) %>% 
  select(condition, go_id = id, description, log_fdr, qvalue, gene_id, gene_ratio, count) %>% 
  filter(qvalue < 0.10) %>%
  group_by(condition) %>% 
  arrange(desc(log_fdr)) %>% 
  slice_head(n = 5)

```



#### Heatmap
```{r}

# Anotações das amostras/condições (Vacinas)
annotation_conditions

heatmap_data = go_filtered
file = "degs_up_down_notimmune_ora_top10"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_log_fdr = heatmap_data %>% 
  select(description, condition, log_fdr)

# Converter para wide table
matrix_data <- dcast(heatmap_data_log_fdr, 
                         description ~ condition, 
                         value.var = "log_fdr", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('description') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_conditions_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(annotation_conditions, by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_condition_complex, by = "condition", all.x = TRUE) %>% 
  arrange(dose) %>% 
  relocate(dose, .before=vaccine) %>% 
  column_to_rownames(var= "condition")  %>% 
  mutate(dose = ifelse(is.na(dose), "NA", dose))

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_conditions_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(dose) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c("...1":previous_vaccination)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() #%>% 
  #scale() #normalizar

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Ordenar colunas
ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  select(type, week, day)


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(0, 3), c("white", "#ed6a5a"))

#Parameters
mat = matrix_data_ordered
width_image = 30
height_image = 30
width = unit(15, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= paste0(file, sep ="_", "Ungrouped_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width= width_image, height=height_image ,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "NES", #Legend
                        cell_fun = function(j, i, x, y, width, height, fill) {
        if(mat[i, j] > 0)
            grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = 8))},
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        # row_split = NULL, #Split row clusters
                        # row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_split = 3, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()

```





### Gene set enrichment analysis (GSEA)

```{r}
#Function to perform GSEA on all conditions
autoGSEA <- function(df, TERM2GENE, geneset_name) {
  resultados <- list()

  conditions <- df$condition %>% 
    unique() %>% 
    as.character()

  for (condition in conditions) {
    # Selecionar DEGs para a condição atual
    degs_condition <- df %>% 
      filter(condition == condition) %>%
      select(genes, log2fold_change) %>%
      distinct() %>%
      mutate(rank = rank(log2fold_change, ties.method = "random")) %>%
      arrange(desc(rank))

    # Garantir que log2fold_change seja um vetor
    gene_list_lf2c <- as.vector(degs_condition$log2fold_change)
    names(gene_list_lf2c) <- degs_condition$genes
    gene_list_lf2c <- na.omit(gene_list_lf2c)
    gene_list_lf2c <- sort(gene_list_lf2c, decreasing = TRUE)

    ############ IMMUNE GO
    auto_gsea <- tryCatch({
      GSEA(geneList = gene_list_lf2c,
           TERM2GENE = TERM2GENE,
           minGSSize = 1,
           maxGSSize = 1000,
           pvalueCutoff = 0.25,
           pAdjustMethod = "BH") %>%
        as.data.frame() %>%
        arrange(qvalue) %>%
        mutate(condition = condicao,
               gsea_enrichment = "ImmuneGO")
    }, error = function(e) {
      return(NULL)
    })

    if (!is.null(auto_gsea)) {
      results[[paste(condition, geneset_name, sep = "_")]] <- auto_gsea
    }
  }
  
  final_result <- bind_rows(results)

  return(final_result)
}
```


```{r}
# Perform GSEA
degs_GSEA <- degs %>%
  select(condition, genes, log2fold_change) %>% 
  autoGSEA(TERM2GENE = TERM2GENE, 
           geneset_name = geneset_name) 

degs_GSEA_ready = degs_GSEA %>% 
  rownames_to_column("delete") %>% 
  select(condition, ID, everything()) %>% 
  clean_names()

#Save
write.csv(degs_GSEA_ready, 
          file = paste0(filename, "_", geneset_name, "_GSEA_ALL_", today(), ".csv"))

view(degs_GSEA_ready)

```





### Shared DEGs Heatmap

Processamento de tabelas Extraindo genes de enriched gene sets

```{r}
#Abrir tabela de GSEA
dados = degs_GSEA_ready

dados_long <- dados %>%
  mutate(core_enrichment = strsplit(core_enrichment, "/")) %>%  # Divida a coluna em listas de genes
  unnest(core_enrichment)  # Transforme a lista em várias linhas

#Filtrar genes por categorias
genes_adaptive <- dados_long %>%
  filter(`Immune system` == "Adaptive")
```

Calculando media, soma etc. por categorias em uma tabela longa

```{r}
dados = DEGs_All_studies_ImmuneGO_GSEA_1_
# Selecione as colunas de interesse
names(dados)

# Calcular a soma dos valores da coluna NES para cada combinação de Vaccine e Immune system
resultado <- dados %>%
  group_by(Vaccine, `Immune system`) %>%
  summarise(`NES sum` = sum(NES))
```

Transformando tabela de DEGs por coluna em tabela longa

```{r}
#Converter tabela de DEGs Up and Down em tabela longa

degs_all = DEGs_All_studies_DEGs_All_Studies
degs_all_long <- pivot_longer(degs_all, cols = everything(), names_to = "Vaccine", values_to = "Genes")

#Salvar
write.csv(degs_all_long, file="degs_all_long.csv")
```

Transformando tabela longa em tabela wide para pca e heatmap

Shared Genes

```{r}
#Selecionar colunas de comparação e valores. Trocar value.var pela variável de interesse ("Shared", "-log(p)")

sharedDEGs_DOWN_fisher_sig = sharedDEGs_DOWN_fisher %>% filter()

matrix_data <- dcast(sharedDEGs_DOWN_fisher, `Cond1` ~ `Cond2`, value.var = "Percentage_Shared_Cond1", fun.aggregate = mean)

#Converter coluna 1 em rownames e excluir
matrix_data$Cond1 = as.factor(matrix_data$Cond1)
rownames(matrix_data) <- matrix_data$Cond1
matrix_data <- matrix_data %>% select(-`Cond1`)

#Arredondar valores para facilitar visualização
matrix_data <- round(matrix_data, 1)
matrix_data <- replace(matrix_data, matrix_data == "NaN", 0)

#Salvar
write.csv(matrix_data, file = "heatmap_sharedDEGs_UP_shared_percent.csv")

#Visualizar
heatmaply(matrix_data, 
          main = "Shared DEGs - UP-DOWN - %identity", 
          cellnote = matrix_data, 
          cellnote_size = 8,
          cellnote_textposition = "middle right",
  scale_fill_gradient_fun = ggplot2::scale_fill_gradient2(
    low = "white", 
    high = "red", 
    midpoint = 1, 
    limits = c(0, 60)
  ))
```

Down vs UP only

```{r}
####### DOWN vs UP only ######## 
filtered_data <- shared_genes_df_mirror %>%
  filter((grepl("UP", Cond1) & grepl("DOWN", Cond2)) | (grepl("DOWN", Cond1) & grepl("UP", Cond2)))

matrix_data <- dcast(filtered_data, `Cond1` ~ `Cond2`, value.var = "Percentage_Shared_Cond1", fun.aggregate = mean)

#Converter coluna 1 em rownames e excluir
rownames(matrix_data) <- matrix_data$`Cond1`
matrix_data <- matrix_data %>% select(-`Cond1`)

#Arredondar valores para facilitar visualização
matrix_data <- round(matrix_data, 1)
matrix_data <- replace(matrix_data, matrix_data == "NaN", 0)

#Arredondar valores para facilitar visualização
matrix_data <- round(matrix_data, 1)

#Salvar
write.csv(matrix_data, file = "heatmap_sharedDEGs_UP_vs_DOWN_shared_percent.csv")

#Visualizar
heatmaply(matrix_data, 
          main = "Shared DEGs - UP vs. DOWN - %identity", 
          cellnote = matrix_data, 
          cellnote_size = 8,
          cellnote_textposition = "middle right",
  scale_fill_gradient_fun = ggplot2::scale_fill_gradient2(
    low = "white", 
    high = "red", 
    midpoint = 1, 
    limits = c(0, 100)
  )
)

```



#### Gráfico de barras

Innate

```{r}
#Definir conjunto de dados
Immune_GO_GSEA$`-log(qvalue)` = as.numeric(Immune_GO_GSEA$`-log(qvalue)`)

# Filtrar as linhas onde a coluna "Gene set" está em valores_desejados
GO_set_innate <- Immune_GO_GSEA %>% filter(`Gene set short` %in% c("INNATE IMMUNE RESPONSE")) %>% mutate(`-log(qvalue)` = as.numeric(`-log(qvalue)`))

GO_set_innate = merge(GO_set_innate, ann_vaccines_complex, by.x = "Vaccine", by.y = "Condition")
esquisser(GO_set_innate)

GO_set_innate_bars = ggplot(GO_set_innate) +
  aes(
    x = Day,
    fill = `Vaccine group`,
    colour = Vaccine,
    weight = NES
  ) +
  geom_bar(position = "dodge") +
  scale_fill_viridis_d(option = "magma", direction = 1) +
  scale_color_viridis_d(option = "magma", direction = 1) +
  labs(y = "-LOG(Q)") +
  theme_minimal() +
  facet_wrap(vars(Dose))

ggsave(GO_set_innate_bars, file = "GO_set_innate_bars.png")

```

```{r}
#Definir conjunto de dados
Immune_GO_GSEA$`-log(qvalue)` = as.numeric(Immune_GO_GSEA$`-log(qvalue)`)

# Filtrar as linhas onde a coluna "Gene set" está em valores_desejados
GO_set_adap <- Immune_GO_GSEA %>% filter(`Gene set short` %in% c("ADAPTIVE IMMUNE RESPONSE")) %>% mutate(`-log(qvalue)` = as.numeric(`-log(qvalue)`))

GO_set_adap = merge(GO_set_adap, ann_vaccines_complex, by.x = "Vaccine", by.y = "Condition")
esquisser(GO_set_adap)

GO_set_adap_bars = ggplot(GO_set_adap) +
  aes(
    x = Day,
    fill = `Vaccine group`,
    colour = Vaccine,
    weight = NES
  ) +
  geom_bar(position = "dodge") +
  scale_fill_viridis_d(option = "magma", direction = 1) +
  scale_color_viridis_d(option = "magma", direction = 1) +
  labs(y = "-LOG(Q)") +
  theme_minimal() +
  facet_wrap(vars(Dose))

ggsave(GO_set_adap_bars, file = "GO_set_adap_bars.png")

```

```{r}
GO_set_humoral <- Immune_GO_GSEA %>% filter(`Gene set short` %in% c("innate immune response", "adaptive immune response", "immunoglobulin mediated immune response", "T cell activation"))



#Definir nome para arquivos gerados
file_bar <- "Immune_GO_general"
fileimage <- paste(file_bar, "_barplot.png", sep = "") 

#Input

barras = 	GO_set_general #Trocar

#Criar gráfico de barras
GO_bars <- ggplot(barras, aes(x = NES, y = `Vaccine`, fill = `-log(qvalue)`)) +
  geom_col(position = "dodge") +
  geom_text(aes(label = round(`-log(qvalue)`, 2)),
            position = position_dodge(width = 0), hjust = 0, vjust = 0, size = 3, angle = 0) +
  theme_minimal() +
  labs(title = fileimage, size = 5,
       x = "NES",
       y = "Vaccine") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 6),
        legend.position = "right") +  # Posição da legenda à direita
  facet_wrap(~`Gene set, short`, ncol = 2) +
  scale_fill_gradient(
    low = "gray90",
    high = "red",# Cor para valores ausentes (NA)
    guide = guide_colorbar(  # Personalizar a legenda de cores
      title = "-log(qvalue)",  # Título da legenda
      title.position = "top",  # Posição do título
      title.hjust = 0.5  # Ajuste de alinhamento do título
    )
  )

GO_bars_bg = GO_bars + theme(
  plot.background = element_rect(fill = "white"),  # Fundo branco
  panel.background = element_rect(fill = "white"),  # Fundo branco
  legend.key = element_rect(fill = "white"),  # Fundo branco para a legenda
  legend.background = element_rect(fill = "white"),  # Fundo branco para a legenda
  legend.title = element_text(color = "black")  # Cor do título da legenda
)

# Salvar
ggsave(GO_bars_bg, file=fileimage, width = 10, height = 5, units = "in") #Trocar

# Display
print(GO_bars_bg)
```

#### Gene L2FC by immune process to heatmap

```{r}
#Padronizar tabela
Immune_allgenes_allprocesses = Immune_allgenes_l2fc %>% 
  rename(Condition = study, Genes = genes) %>% filter(Vaccine != "NA")

#General
Immune_GO_general_innate = Immune_allgenes_allprocesses %>% 
  filter(Process == "innate immune response") %>% 
  filter(Condition != "NA")

Immune_GO_general_adaptive = Immune_allgenes_allprocesses %>% 
  rename(Condition = study, Genes = genes) %>% 
  filter(Process == "adaptive immune response",
         Condition != "NA"))

\#Humoral
Immune_GO_humoral_humoral = Immune_GO_humoral %>% 
  rename(Condition = study, Genes = genes) %>% 
  filter(Process == "humoral immune response",
         Condition != "NA"))

Immune_GO_humoral_IgMediated = Immune_GO_humoral %>% 
  rename(Condition = study, Genes = genes) %>% 
  filter(Process == "immunoglobulin mediated immune response", 
         Condition != "NA"))

#Cellular
Immune_GO_adaptive_cellular_tcellactivation = Immune_GO_cellular %>% rename(Condition = study, Genes = genes) %>% filter(Process == "T cell activation") %>% filter(Condition != "NA")
```

#### Criando matriz para heatmap

```{r}
#Input
heatmap_data = Immune_GO_general_adaptive
file= "Immune_GO_general_adaptive"

#Excluir linhas com Target ou Source == NA
heatmap_data = filter(heatmap_data, Vaccine != "NA")

#Selecionar colunas de comparação e valores. Trocar value.var pela variável de interesse
names(heatmap_data)
matrix_data <- dcast(heatmap_data, `Genes` ~ `Condition`, value.var = "log2FoldChange", fun.aggregate = mean)

#Converter coluna 1 em rownames e excluir
rownames(matrix_data) <- matrix_data$`Genes`
matrix_data <- matrix_data %>% select(-`Genes`)

#Arredondar valores para facilitar visualização
matrix_data <- round(matrix_data, 1)
matrix_data <- replace(matrix_data, matrix_data == "NaN", 0)


#Definir nome para arquivo geral
write.csv(matrix_data, file=paste(file, "heatmap.csv",sep="_")) 

```

#### pHeatmap

```{r}

#Definir matriz de dados
matrix_data = as.matrix(matrix_data)

cellmarker_studies

# Criar tabela de anotações para as vacinas (só rodar uma vez)
# rownames(ann_vaccines) = rownames(t(matrix_data)) # Obter condições e transpor para linhas
# ann_vaccines = select(ann_vaccines, -ann_vaccines)
#ann_vaccines = data.frame(ann_vaccines) # Converter em dataframe

# Agrupamento
# ann_vaccines$Vaccine = NA
# ann_vaccines$Dose= NA
# ann_vaccines$Day = NA
# #Anotar manualmente usando editData
# ann_vaccines = editData(ann_vaccines) 

#Cores

# # Defina os níveis do fator Vaccine para corresponderem às cores
# ann_vaccines$Vaccine <- factor(ann_vaccines$Vaccine)
# ann_vaccines$Dose = factor(ann_vaccines$Dose)
# ann_vaccines$Day <- factor(ann_vaccines$Day)

#Defina a ordem das anotações no gráfico
# ann_vaccines = select(ann_vaccines, Day, Dose, Vaccine)

# # Defina as cores correspondentes em annotation_colors
# ann_colors <- list(
#   Vaccine = c(
#   "BBIBP" = "#F46D43",
#   "BNT" = "#0096C7",
#   "ChAd" = "#99d98c",
#   "ChAd-BNT" = "#76c893",
#   "ZF2001" = "#ffd500"),
#   Dose = c(
#     "1" = "#ffea00",
#     "2" = "#ff7b00",
#     "3" = "#e01e37"
#   ),
#   Day = c(
#     "01" = "#a5ffd6",
#     "03" = "#99d98c",
#     "06" = "#00bbf9",
#     "07" = "#4cc9f0",
#     "14" = "#7b2cbf",
#     "28" = "#b5179e"))


# Defina os valores mínimos e máximos para o espectro de cores
valores_minimos <- -1  #  Substitua pelo valor mínimo desejado
valores_maximos <- 3   # Substitua pelo valor máximo desejado

# min(matrix_data)
# max(matrix_data)



# Crie o gráfico de calor com configurações personalizadas
p = pheatmap::pheatmap(
matrix_data,
annotation_colors = ann_colors,
annotation_col = ann_vaccines,
cutree_cols = 5,
cutree_rows = 7,
width = 100,  # Largura da figura (ajuste conforme necessário)
height = 1000,  # Altura da figura (ajuste conforme necessário)
fontsize_row = 2,  # Tamanho da fonte para rótulos de linhas
fontsize_col = 3,  # Tamanho da fonte para rótulos de colunas
cluster_rows = TRUE,  # Impede o agrupamento de linhas
cluster_cols = F,  # Impede o agrupamento de colunas
main = "Immune_GO_adaptive_cellular_tcellactivation",
angle_col = 90,  # Rotação dos rótulos de coluna (0 para horizontal)
cellheight = 2,  # Altura das células (ajuste conforme necessário)
legend = TRUE,
border = TRUE,
border_color = "black",
show_colnames = TRUE,
show_rownames = TRUE,
color = colorRampPalette(c("blue", "white", "red"))(100),  # Esquema de cores
breaks = c(seq(valores_minimos, -0.001, length.out = 50), seq(0.001, valores_maximos, length.out = 50)))


# Salve o gráfico como um arquivo PNG
png(file= fileimage, width=2000, height=2000, res=315, pointsize=10)
print(p)  # Substitua 'p' pelo objeto 'pheatmap'
dev.off()  # Finalize a gravação do arquivo


```

#### Gráfico de barras

```{r}

#Definir conjunto de dados
gene_set = ImmuneGO_Annotated_Genes

#Definir nome para arquivos gerados
file <- "Immune_GO"
fileimage <- paste(file, "_barplot.png", sep = "") 


#Separar por vacina
# BBIBP = filter(gene_set, Vaccine == "BBIBP")
# ZF2001 = filter(gene_set, Vaccine == "ZF2001")
# BNT = filter(gene_set, Vaccine == "BNT")
# ChAd = filter(gene_set, Vaccine == "ChAd")

#Input
barras = gene_set #Trocar

#Criar gráfico de barras
genes_bars <- ggplot(barras, aes(x = Genes, y = log2FoldChange, fill = Condition)) +
  geom_col(position = "dodge") +
  geom_text(aes(label = round(log2FoldChange, 1)),
            position = position_dodge(width = 0), hjust = 0, vjust = 0, size = 1.5, angle = 0) +
  theme_minimal() +
  labs(title = "Innate immune response associated genes - BBIBP", size = 5, #Trocar
       x = "Gene",
       y = "L2FC") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 6),
        legend.position = "none") + 
  facet_wrap(~Condition, ncol = 1)

genes_bars

ggsave(genes_bars, file=fileimage, width = 8, height = 16, units = "in") #Trocar
```




```{r}
all_degs_p_05_vac_infected_29_11_23 %>% 
  filter(condition == "BNT (V2, D1)") %>% 
  select(genes) %>% 
  write.csv(file = "degs_BNT_V2_D1.csv")
```

